man/migrateddumbfs.8
.TH "MIGRATEDDUMBFS" "8" "January 11, 2013" "1.0" "ddumbfs"
.SH NAME
migrateddumbfs \- command line tools to resize a ddumbfs filesystem
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructeredText.
.
.SH SYNOPSIS
.sp
\fBmigrateddumbfs\fP [options] \fIsource\-parent\-directory\fP \fIdestination\-parent\-directory\fP
.SH DESCRIPTION
.sp
\fBmigrateddumbfs\fP used to resize a ddumbfs filesystem.
.sp
The \fIsource\-parent\-directory\fP and \fIdestination\-parent\-directory\fP are directory containing
the \fIddfs.cfg\fP file and \fIddfsroot\fP directory. \fIdestination\-parent\-directory\fP must exists,
must be empty, must have been created with the same \fIblock_size\fP and \fIhash\fP and
must be big enough to receive the data of the source. The \fIblock file\fP is not modified
and not copied. You have to manually copy/move data of the \fIblock file\fP.
.SH OPTIONS
.INDENT 0.0
.TP
.B \-h, \-\-help
Show help message and exit.
.UNINDENT
.INDENT 0.0
.TP
.B \-v, \-\-verbose
Be more verbose.
.UNINDENT
.INDENT 0.0
.TP
.B \-p, \-\-progress
Display progress second by second of some part of the file system check.
.UNINDENT
.INDENT 0.0
.TP
.B \-f, \-\-force
bypass some guardrail
.UNINDENT
.INDENT 0.0
.TP
.B \-m, \-\-migrate
Default option for now, use it to avoid compatibility problems with
future release that could have more features
.UNINDENT
.SH EXAMPLES
.sp
Create a new ddumbfs filesystem and migrate data to it.:
.sp
.nf
.ft C
fsckddumbfs \-r /data/ddsource
mkddumbfs \-s 100G /data/dddestination
migrateddumbfs \-m /data/ddsource /data/ddestination
rm /data/dddestination/ddfsblocks
mv /data/ddsource/ddfsblocks /data/dddestination/ddfsblocks
.ft P
.fi
.SH SEE ALSO
.sp
\fIddumbfs(1)\fP, \fImkddumbfs(8)\fP, \fIcpddumbfs(1)\fP
.SH AUTHOR
.sp
Alain Spineux <\fI\%alain.spineux@gmail.com\fP>
.SH AUTHOR
Alain Spineux
.SH COPYRIGHT
2011, Alain Spineux
.\" Generated by docutils manpage writer.
.\" 
.


doc/Changelog


2013-01-12 ddumbfs-1.1

    fsckddumbfs:
      - option pack to pack the block file and be able to downsize the filesystem using migrateddumbfs

    all:
        - added .deb generation
        - improved .rpm generation

    ddumbfs:
        - direct_io default is to be enable when block file is a block device
        - no more align option
        
    ddfslib:
        - bug: ddfs_init() was not repairing the blockfile MAGIC

    bits:
     - optimized and added more testing for quick_functions
        

2012-11-23 ddumbfs-1.0b21

    all:
      replaced "struct ddfs_ctx ddfs" by "struct ddfs_ctx *ddfs" to be able to 
      mount multiple filesystem, required by new migrateddumbfs 
      
    migrateddumbfs:
      - new tools to resize ddumbfs (migrate and existing fs into an fs of different size)
        
    ddumbfs:
      - don't start if .autofsck cannot be locked.

    fsckddumbfs:
      - bug: fsckddumbfs was unable to recreate an index.


2012-11-21 ddumbfs-1.0b20

    all:
      - changed logging to append mode to avoid a rollout every time ddumbfs or 
      any ddfs utils is started and finnaly loose log.  
      - don't check for libaio anymore in autoconf
      
    ddumbfs:
      - added ddumb_statistic.inode_counter and display ddumb_statistic.fh_counter

2012-11-17 ddumbfs-1.0b19
    
    bits:
      - added bit_array_cmp_count() 

    ddumbfs:
      - use bit_array_cmp_count() to reduce complexity 

    fsckddumbfs:
      - all block reading and hashing that were using libaio are no using 
      multi-threading to improve performance.  

2012-11-07 ddumbfs-1.0b18
    
    ddumbfs:
      - bug: ddfs_save_usedblocks() was not SYNCING bfile and ifile, now it does
      - added c_auto_buffer_flush=60s to force flush of pending buffer and 
        c_auto_sync=120s to force sync of bfile and ifile. Can be tuned in ddfs.cfg
      
    fsckddumbfs:
      - bug fix that generate SIGSEV
      

2012-11-05 ddumbfs-1.0b17

    ddumbfs:
      - now ddumb_fsync() sync the bfile and the ifile too

    queryddumbfs:
      - added option "--force" to run even when filesystem is already mounted
      - it was not checking reslut of ddfs_read_full_block(), now it does


2012-10-31 ddumbfs-1.0b16

    ddumbfs:
      - lot of cosmetic change and renamed some variables and function 
      - fixed a bug in block_wait() 

2012-10-29 ddumbfs-1.0b15

    all:
      - close and closer to the 1.0 :-)

    ddumbfs:
      - bug: Access to file in .ddumbfs was locking the thread because these
      file dont have fh->lock and ddumbfs_read() was loccking fh->lockfor any
      file


2012-10-26 ddumbfs-1.0b14

    all:
      - in Makefile.am, replaced AM_LDFLAGS into LDADD, because of ubuntu 12.04

    ddumbfs:
      - flush "long living" buffer at regular interval 
      
    ddfschkrep:
      - check only regular files

    fsckddumbfs:
      - check only regular files


2011-11-21 ddumbfs-1.0b13

    all:
      - use bits mask and bits shift instead of mod, div and mul to get gap,
      and block_boundary from an offset.
    
    ddfslib      
      - improved ddfs_convert_addr() and ddfs_get_node_addr()

    ddumbfs:
      - the lock_index was not working because mlock() was done before
      the fork() done in fuse_main(). Now mlock is done in ddumbfs_init() 

2011-11-12 ddumbfs-1.0b12

    ddumbfs:
      - handle the unlikely race condition on block write using a list of block 
      being written. Any read check this list and must wait if the block is in the list
      - delayed error was wrongly reported 
      
    testddumbfs:
      - added operation 'B' to test for the 'race condition' above but this is 
      useless because FUSE cache the data and never issue a read.
      
      
2011-11-00 ddumbfs-1.0b11
    ddumbfs:
      - write requests are queued in a list and writers pick them up.
      - some tuning when waiting for a "REANDONLY" fh to be written (in _ddumb_write()
      and do_truncate())

    !! released soon, mostly an alpha  

2011-10-25 ddumbfs-1.0b10

    mkddumbfs:
      - bug: 128K block is the default
      - bug: c_addr_size was wrongly calculated, 1 bit too small and then
      because of the use of octet, 1/8 wrong !
      
    queryddumbfs:
      - NEW tool to query and search the ddumbfs index and files about nodes, 
      hash and blocks.

2011-10-25 ddumbfs-1.0b9
    xlog:
      - was rolling out twice at startup when not in "append" mode

    ddumbfs:
      - removed xstat->ref because was xstat->fhs_n similar 
      - bug: in writer_pool_handle() "fh_dst->pool_loaded=1" moved before xstat_subscribe() 
      - added option "reclaim" to handle auto reclaim when disk usage > reclaim%
       

2011-10-23 ddumbfs-1.0b8
    xlog:
      - added filename alias, to handles file like 'stderr' 
      - error handling
    ddumbfs:
      - close file after read: "/proc/cpuinfo"
      - reclaim() handle non regular file like (sockets, fifos, block
      & char devices) better
      - symblink set the uid and gid appropriately
      - pjd succeed 
      - added option [no]fuse_default that insert: "-ouse_ino,readdir_ino,
         ...default_permissions,allow_other,big_writes,max_read=131072,
         ...max_write=131072,negative_timeout=0,entry_timeout=0,attr_timeout=0
      - "fuse_default" can be over-written
      - bug in xstat_compare() was generating "xstat_release: xstat not found, continue"
      use long long int insead of int
      - complete rewrite of the writer pool. Writers are in a list, the requester
      iterates the list until a free writer is found, upload data in the ddumb_fh
      and signal the thread.

    ddfslib:
      - close file after read: "ddfs.cfg"
    mkddumbfs:
      - default block size is now 128k
    cpddumbfs:
      - added some code to "download" file even if bad block are found.
      
2011-10-10 ddumbfs-1.0b7
    ddumbfs:
      - added late block allocation to reduce self-fragmentation and increase

src/migrateddumbfs.c
/**
 * @migrateddumbfs.c
 * @author  Alain Spineux <alain.spineux@gmail.com>
 * @version 0.0
 *
 * @section LICENSE
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of
 * the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details at
 * http://www.gnu.org/copyleft/gpl.html
 *
 * @section DESCRIPTION
 *
 * migrateddumbfs allows to resize a ddumbfs filesytem by migrating
 * data data between thow fs of different size
 *
 */

#define _XOPEN_SOURCE 500
#define _LARGEFILE64_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <ftw.h>
#include <fcntl.h>
#include <assert.h>

#include "ddfslib.h"

#define MAX_SIZE 0xFFFFFFFFFFFFFFFFll;

int verbose_flag=0;
int check_integrity=0;
int lock_index_flag=0;
int force_flag=0;
int progress_flag=0;

static struct option long_options[] =
{
       /* These options set a flag. */
       {"help",          no_argument,       0, 'h'},
       {"verbose",       no_argument,       0, 'v'},
       {"force",         no_argument,       0, 'f'},
       {"progress",      no_argument,       0, 'p'},

       {0, 0, 0, 0}
};

void usage()
{
    printf("Usage: migrateddumbfs [options] parent_source parent_destination\n"
            "\n    copy from and to an offline ddumbfs filesystem\n\nOptions\n"
            "  -h, --help            show this help message and exit\n"
            "  -v, --verbose         display addresses and hashes from file (download only)\n"
            "  -p, --progress        display progression\n"
            "  -f, --force           download a file even with a corrupted MAGIC in the header\n"
            "\n  migrate nodes of the index and the files from source to destination\n"
            "  directory. Destination must exists. Blockfile will not be migrated. "
    		"  Its format don't need to be changed for a filesystem resize. You"
            "  must manually edit ddfs.cfg file to link the blockfile to the old one.\n"
            "\nSamples:\n"
            "  migratedumbfs /data/ddumbfs /data/ddumbfs.new\n"
    );
}

int ddfs_index_migrate(struct ddfs_ctx *src, struct ddfs_ctx *dst)
{
    nodeidx node_idx, dst_idx;

    long long int start=now();
    long long int last=start;

    ddfs=src;
    for (node_idx=dst_idx=0; node_idx<src->c_node_count; node_idx++)
    {
        unsigned char *node=src->nodes+(node_idx*src->c_node_size);
        unsigned char *hash=node+src->c_addr_size;

        blockaddr addr=ddfs_get_node_addr(node);

        if (progress_flag && node_idx%10000==0)
        {	// display progress
            long long int end=now();
            if (end-last>NOW_PER_SEC)
            {
                last=end;
                printf("migrating index nodes: %.1f%% in %llds\r", node_idx*100.0/src->c_node_count, (end-start)/NOW_PER_SEC);
                fflush(stdout);
            }
        }

        if (addr!=0)
        {
        	ddfs=dst;
            nodeidx cidx=ddfs_hash2idx(hash);
            if (cidx<dst_idx) cidx=dst_idx;
            dst_idx=cidx+1;

            unsigned char *node_dst=dst->nodes+(cidx*dst->c_node_size);
            unsigned char *hash_dst=node_dst+dst->c_addr_size;

			ddfs_convert_addr(addr, node_dst);
			memcpy(hash_dst, hash, dst->c_hash_size);
        }
    }

    // copy ba_usedblocks
    printf("migrate bit list\n");
    node_idx=bit_array_search_first_set(&src->ba_usedblocks, 0);
    while (node_idx>=0)
    {
        bit_array_set(&dst->ba_usedblocks, node_idx);
        node_idx=bit_array_search_first_set(&src->ba_usedblocks, node_idx+1);
    }

    return 0;
}

struct ddfs_ctx ddfs_src;
struct ddfs_ctx ddfs_dst;
long long int ddfs_tree_migrate_te_file_count;
long long int ddfs_tree_migrate_te_errors;
long long int ddfs_tree_migrate_te_last;

int ddfs_tree_migrate(const char *fpath, const struct stat *sb, int typeflag, struct FTW *ftwbuf)
{
    int res, len;
    char filename_dst[FILENAME_MAX];
    unsigned char node_src[NODE_SIZE];
    unsigned char node_dst[NODE_SIZE];
    unsigned char *hash_src=node_src+ddfs_src.c_addr_size;
    unsigned char *hash_dst=node_dst+ddfs_dst.c_addr_size;
    FILE *file_dst=NULL;
    FILE *file_src=NULL;
    uint64_t size;

    if (typeflag!=FTW_F || ! S_ISREG(sb->st_mode)) return 0;

    if (0==strncmp(fpath+ddfs_src.rdir_len, SPECIAL_DIR, ddfs_src.special_dir_len))
    {   // skip SPECIAL_DIR
        return 0;
    }

    ddfs_tree_migrate_te_file_count++;
    if (verbose_flag)
    {
        long long int end=now();
        if (end-ddfs_tree_migrate_te_last>NOW_PER_SEC)
        {
            ddfs_tree_migrate_te_last=end;
            printf("Read file: %lld\r", ddfs_tree_migrate_te_file_count);
            fflush(stdout);
        }
    }

    file_src=fopen(fpath, "rb");
    if (file_src==NULL)
    {
        ddfs_tree_migrate_te_errors++;
        printf("cannot open: %s (%s)\n", fpath, strerror(errno));
        return 0;
    }

    // be careful, file_header_set_conv return c_file_header_size for empty file
    len=file_header_set_conv(fileno(file_src), &size);
    if (len!=ddfs_src.c_file_header_size)
    {
        ddfs_tree_migrate_te_errors++;
        if (verbose_flag)
        {
            if (len==-1) printf("cannot read: %s (%s)\n", fpath, strerror(errno));
            else printf("corrupted header: %s\n", fpath);
        }
        goto END;
    }

    res=fseek(file_src, ddfs_src.c_file_header_size, SEEK_SET);
    if (res==-1)
    {
        ddfs_tree_migrate_te_errors++;
        printf("cannot seek at first block: %s (%s)\n", fpath, strerror(errno));
        goto END;
    }

    // Destination
    snprintf(filename_dst, FILENAME_MAX, "%s/%s", ddfs_dst.rdir, fpath+ddfs_src.rdir_len+1);

//    fprintf(stderr, "src: %s\n", fpath);
//    fprintf(stderr, "dst: %s\n", filename_dst);

    file_dst=fopen(filename_dst, "wb");
    if (file_dst==NULL)
    {
        ddfs_tree_migrate_te_errors++;
        printf("cannot open: %s (%s)\n", filename_dst, strerror(errno));
        goto END;
    }

    res=fseek(file_dst, ddfs_dst.c_file_header_size, SEEK_SET);
    if (res==-1)
    {
        ddfs_tree_migrate_te_errors++;
        printf("cannot seek at first block: %s (%s)\n", filename_dst, strerror(errno));
        goto END;
    }


    long long int i=0;

    len=fread(node_src, 1, ddfs_src.c_node_size, file_src);
    while (len==ddfs_src.c_node_size)
    {
		i++;
    	if (ddfs_src.c_node_size==ddfs_dst.c_node_size)
    	{
			res=fwrite(node_src, 1, ddfs_dst.c_node_size, file_dst);
    	}
    	else
    	{
    		ddfs=&ddfs_src;
			blockaddr addr=ddfs_get_node_addr(node_src);
    		ddfs=&ddfs_dst;
			ddfs_convert_addr(addr, node_dst);
			memcpy(hash_dst, hash_src, ddfs_src.c_hash_size);
			res=fwrite(node_dst, 1, ddfs_dst.c_node_size, file_dst);
			if (res!=ddfs->c_node_size)
			{
		        ddfs_tree_migrate_te_errors++;
				fprintf(stderr, "error writing file: %s\n", filename_dst);
		        goto END;
			}
    		ddfs=&ddfs_src;
    	}

		len=fread(node_src, 1, ddfs_src.c_node_size, file_src);
    }

    if (len!=0)
    {
        ddfs_tree_migrate_te_errors++;
		if (len==-1) printf("cannot read file: %s (%s)\n", fpath, strerror(errno));
		else printf("incomplete node in file: %s %d\n", fpath, len);
        goto END;
    }

    if (i!=(size+ddfs_src.c_block_size-1)/ddfs_src.c_block_size)
    {   // size mismatch
        ddfs_tree_migrate_te_errors++;
        printf("size mismatch in file: %s\n", fpath);
        goto END;
    }

    // write size
    ddfs=&ddfs_dst;
    fflush(file_dst);
    len=file_header_get_conv(fileno(file_dst), size);
    if (len==-1)
    {
        perror(filename_dst);
        goto END;
    }
    else if (len!=ddfs_dst.c_file_header_size)
    {
        fprintf(stderr, "cannot write file header\n");
        goto END;
    }

END:
	ddfs=&ddfs_src;
    if (file_src!=NULL) fclose(file_src);
    if (file_dst!=NULL) fclose(file_dst);
    return 0;
}


struct ddfs_ctx ddfs_src;
struct ddfs_ctx ddfs_dst;

int main(int argc, char *argv[])
{
    int res=0;
    int c;
    long long int start, end;

    while (1)
    {
        // getopt_long stores the option index here.
        int option_index = 0;

        c=getopt_long(argc, argv, "hvfp", long_options, &option_index);

        // Detect the end of the options.
        if (c==-1) break;

        switch (c)
        {

            case 'h':
                usage();
                return 0;
                break;

            case 'v':
                verbose_flag=1;
                break;

            case 'p':
                progress_flag=1;
                break;

            case 'f':
                force_flag=1;
                break;

            case '?':
                // getopt_long already printed an error message.
                break;

            default:
                abort();
        }
    }

    if (optind+1>=argc)
    {
        fprintf(stderr, "missing source or destination parent directory %d %d!\n", optind, argc );
        return 1;
    }

    char pdir_src[FILENAME_MAX];
    char pdir_dst[FILENAME_MAX];

    if (ddfs_find_parent(argv[optind], pdir_src)!=0)
    {
        fprintf(stderr, "Source directory not found !\n");
        return 1;
    }

    if (ddfs_find_parent(argv[optind+1], pdir_dst)!=0)
    {
        fprintf(stderr, "Destination directory not found: %s\n", argv[optind+1]);
        return 1;
    }

    if (is_mounted(pdir_src))
    {
        fprintf(stderr, "source filesystem already mounted: %s\n", pdir_src);
        return 1;
    }

    if (is_mounted(pdir_dst))
    {
        fprintf(stderr, "destination filesystem already mounted: %s\n", pdir_dst);
        return 1;
    }

    ddfs=&ddfs_dst;
    printf("mounting destination: %s\n", pdir_dst);
    if (ddfs_loadcfg(pdir_dst, stderr))
    {
        return 1;
    }

    if (ddfs_init(DDFS_NOFORCE, DDFS_NOREBUILD, DDFS_DIRECTIOAUTO, DDFS_NOLOCKINDEX, NULL))
    {
        return 1;
    }

    ddfs=&ddfs_src;
    printf("mounting source: %s\n", pdir_src);
    if (ddfs_loadcfg(pdir_src, NULL))
    {
        return 1;
    }

    if (ddfs_init(DDFS_NOFORCE, DDFS_NOREBUILD, DDFS_DIRECTIOAUTO, DDFS_NOLOCKINDEX, NULL))
    {
        return 1;
    }

    printf("check source and destination compatibility\n");
    if (ddfs_src.c_block_size!=ddfs_dst.c_block_size)
    {
        printf("blocksize mismatch !\n");
        res=1;
        goto END;
    }

    if (0!=strcmp(ddfs_src.c_hash, ddfs_dst.c_hash))
    {
        printf("HASH mismatch !\n");
        res=1;
        goto END;
    }
    // compare used block
    long long int src_block_count, src_block_free;
    long long int dst_block_count, dst_block_free;

    bit_array_count(&ddfs_src.ba_usedblocks, &src_block_count, &src_block_free);
    bit_array_count(&ddfs_dst.ba_usedblocks, &dst_block_count, &dst_block_free);

    if (dst_block_count!=DDFS_LAST_RESERVED_BLOCK+1)
    {
        printf("Destination must be empty, please format destination !\n");
        res=1;
        goto END;
    }

    if (dst_block_free<src_block_count)
    {
        printf("Not enough free blocks in destination %lld < %lld!\n", dst_block_free, src_block_count);
        res=1;
        goto END;
    }

    printf("== migrating index\n");
    ddfs_index_migrate(&ddfs_src, &ddfs_dst);

    printf("migrating free block list\n");


    printf("== migrating file\n");
    start=now();
    ddfs_tree_migrate_te_file_count=0;
    ddfs_tree_migrate_te_last=start;
    res=nftw(ddfs_src.rdir, ddfs_tree_migrate, 10, FTW_MOUNT | FTW_PHYS);
    end=now();
    printf("== Migrated %lld files in %.1fs\n", ddfs_tree_migrate_te_file_count, (end-start)*1.0/NOW_PER_SEC);


    printf("don't forget to handle the blockfile manually\n");
END:
	ddfs=&ddfs_src;
    ddfs_close();
    ddfs=&ddfs_dst;
    ddfs_close();

    return res;

}
